pkg/rhcos/openstack.go

```diff
@@ -25,7 +34,114 @@ func OpenStack(ctx context.Context) (string, error) {
 		return "", err
 	}
 
-	return base.ResolveReference(relOpenStack).String(), nil
+	// Attach uncompressed sha256 checksum to the URL
+	checkSuffix := "?sha256=" + meta.Images.OpenStack.UncompressedSHA256
+	baseURL := base.ResolveReference(relOpenStack).String() + checkSuffix
+
+	// Check that we have generated a valid URL
+	_, err = url.ParseRequestURI(baseURL)
+	if err != nil {
+		return "", err
+	}
+
+	return baseURL, nil
+}
+
+func getCacheDir() (string, error) {
+	// TODO(mfedosin): os.UserCacheDir was added in golang 1.11, but since we still use 1.10
+	// we can't use it directly. Uncomment this when we start using golang 1.11+ for building
+	// the binary.
+	//defaultCacheDir, err := os.UserCacheDir()
+
+	defaultCacheDir := configdir.New("openshift-installer", "images-cache").QueryCacheFolder()
+
+	defaultCacheDir.MkdirAll()
+
+	return defaultCacheDir.Path, nil
+}
+
+func validateSHA256Checksum(filepath, checksum string) error {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+
+	if checksum != fmt.Sprintf("%x", h.Sum(nil)) {
+		logrus.Fatal("Image sha256 checksum is invalid.")
+		defer os.Remove(filepath)
+		return errors.Errorf("image sha256 checksum is invalid")
+	}
+
+	return nil
+}
+
+// DownloadOpenStackImageFile obtains a file from a given URL, puts it in the cache
+// and returns the local file path
+func DownloadOpenStackImageFile(baseURL string) (string, error) {
+	logrus.Infof("Obtaining RHCOS image file from '%v'", baseURL)
+
+	// Convert the given URL into a file name using md5 algorithm
+	imageFileName := fmt.Sprintf("%x.img", md5.Sum([]byte(baseURL)))
+
+	// If the file has already been cached, return its path
+	cacheDir, err := getCacheDir()
+	if err != nil {
+		return "", err
+	}
+	filePath := filepath.Join(cacheDir, imageFileName)
+
+	_, err = os.Stat(filePath)
+	if !os.IsNotExist(err) {
+		logrus.Infof("The file was found in cache. Reusing...")
+		return filePath, nil
+	}
+
+	logrus.Debugf("Caching image file in '%v'", filePath)
+
+	// Create the file
+	out, err := os.Create(filePath)
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	// Get the data
+	resp, err := http.Get(baseURL)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.Errorf("bad status: %s", resp.Status)
+	}
+
+	// Writer the body to file
+	_, err = io.Copy(out, resp.Body)
```

Comment:
Storing the whole image in memory here is unfortunate.  We should be able to stream the body and checksum at the same time.

Reply:
Done